File: test/test_remote_connection.py
#!/usr/bin/env python3
"""
Test script for connecting to remote readsb/Ultrafeeder MCP server
"""

import asyncio
import json
import subprocess
import sys
from pathlib import Path

# Add parent directory to path for imports
sys.path.insert(0, str(Path(__file__).parent.parent))
from shared_utils import get_remote_host_and_port  # noqa: E402


async def test_remote_mcp_server(remote_host: str, remote_port: int = 8080):
    """Test the MCP server with remote readsb/Ultrafeeder"""

    print(f"Testing MCP Server Connection to {remote_host}:{remote_port}...")
    print("=" * 60)

    # Construct the base URL for the remote feeder
    base_url = f"http://{remote_host}:{remote_port}"

    # Start the MCP server process
    server_path = Path(__file__).parent.parent / "readsb_mcp_server.py"

    server_process = subprocess.Popen(
        [
            "uv",
            "run",
            "python",
            server_path,
            "--base-url",
            base_url,
        ],
        stdin=subprocess.PIPE,
        stdout=subprocess.PIPE,
        stderr=subprocess.PIPE,
        text=True,
    )
    if not server_process or not server_process.stdin or not server_process.stdout:
        raise RuntimeError("Failed to start MCP server process")

    try:
        # Send initialization request
        init_request = {
            "jsonrpc": "2.0",
            "id": 1,
            "method": "initialize",
            "params": {
                "protocolVersion": "2024-11-05",
                "capabilities": {"roots": {"listChanged": True}, "sampling": {}},
                "clientInfo": {"name": "test-client", "version": "1.0.0"},
            },
        }

        print("Sending initialization request...")
        server_process.stdin.write(json.dumps(init_request) + "\n")
        server_process.stdin.flush()

        # Read initialization response
        response_line = server_process.stdout.readline()
        if response_line:
            response = json.loads(response_line)
            print(f"Initialization response: {response.get('result', {}).get('serverInfo', {}).get('name', 'Unknown')}")

        # Send initialized notification
        initialized_notification = {"jsonrpc": "2.0", "method": "notifications/initialized", "params": {}}

        server_process.stdin.write(json.dumps(initialized_notification) + "\n")
        server_process.stdin.flush()

        # Test list tools
        list_tools_request = {"jsonrpc": "2.0", "id": 2, "method": "tools/list"}

        print("Requesting tools list...")
        server_process.stdin.write(json.dumps(list_tools_request) + "\n")
        server_process.stdin.flush()

        tools_response = server_process.stdout.readline()
        if tools_response:
            tools_data = json.loads(tools_response)
            tools = tools_data.get("result", {}).get("tools", [])
            print(f"Available tools: {[tool['name'] for tool in tools]}")

        # Test receiver stats first (less likely to fail)
        print("\nTesting get_receiver_stats...")
        stats_request = {
            "jsonrpc": "2.0",
            "id": 3,
            "method": "tools/call",
            "params": {"name": "get_receiver_stats", "arguments": {"format": "summary"}},
        }

        server_process.stdin.write(json.dumps(stats_request) + "\n")
        server_process.stdin.flush()

        stats_response = server_process.stdout.readline()
        if stats_response:
            stats_data = json.loads(stats_response)
            if "result" in stats_data:
                print("Receiver stats successful!")
                print(f"Response: {stats_data['result']['content'][0]['text']}")
            else:
                print(f"Receiver stats error: {stats_data.get('error', 'Unknown error')}")

        # Test aircraft data
        print("\nTesting get_closest_aircraft...")
        aircraft_request = {
            "jsonrpc": "2.0",
            "id": 4,
            "method": "tools/call",
            "params": {"name": "get_closest_aircraft", "arguments": {"count": 3}},
        }

        server_process.stdin.write(json.dumps(aircraft_request) + "\n")
        server_process.stdin.flush()

        aircraft_response = server_process.stdout.readline()
        if aircraft_response:
            aircraft_data = json.loads(aircraft_response)
            if "result" in aircraft_data:
                print("Aircraft data successful!")
                response_text = aircraft_data["result"]["content"][0]["text"]
                print(f"Response: {response_text[:300]}...")
            else:
                print(f"Aircraft data error: {aircraft_data.get('error', 'Unknown error')}")

        # Test directional query
        print("\nTesting get_aircraft_by_direction (east)...")
        direction_request = {
            "jsonrpc": "2.0",
            "id": 5,
            "method": "tools/call",
            "params": {"name": "get_aircraft_by_direction", "arguments": {"direction": "east", "count": 2}},
        }

        server_process.stdin.write(json.dumps(direction_request) + "\n")
        server_process.stdin.flush()

        direction_response = server_process.stdout.readline()
        if direction_response:
            direction_data = json.loads(direction_response)
            if "result" in direction_data:
                print("Directional query successful!")
                response_text = direction_data["result"]["content"][0]["text"]
                print(f"Response: {response_text[:300]}...")
            else:
                print(f"Directional query error: {direction_data.get('error', 'Unknown error')}")

    except Exception as e:
        print(f"Error during testing: {e}")
        print("This might indicate network connectivity issues or incorrect host/port")

    finally:
        # Clean up
        server_process.terminate()
        server_process.wait()
        print("\nMCP server test completed.")


def main():
    """Main entry point with user input"""
    print("Remote readsb MCP Server Test")
    print("=" * 30)

    # Get remote host and port from user
    remote_host, remote_port = get_remote_host_and_port()

    print(f"\nTesting connection to {remote_host}:{remote_port}")
    print("Make sure your readsb/Ultrafeeder is running and accessible")

    # Run the test
    asyncio.run(test_remote_mcp_server(remote_host, remote_port))


if __name__ == "__main__":
    main()

File: test/test_remote_connection.ts
#!/usr/bin/env node
/**
 * Basic connection test for remote readsb/Ultrafeeder
 */

import { spawn } from 'child_process';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { getRemoteHostAndPort } from '../src/shared_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

export async function testRemoteMcpServer(remoteHost: string, remotePort: number = 8080): Promise<void> {
  console.log('Remote readsb MCP Server Test');
  console.log('='.repeat(30));

  const baseUrl = `http://${remoteHost}:${remotePort}`;
  console.log(`Testing connection to: ${baseUrl}`);

  // Build the server script path
  const serverPath = path.join(__dirname, '..', 'readsb_mcp_server.js');

  console.log('Starting MCP server process...');

  // Start the server process
  const serverProcess = spawn('node', [serverPath, '--base-url', baseUrl], {
    stdio: ['pipe', 'pipe', 'pipe'],
    cwd: path.join(__dirname, '..'),
  });

  if (!serverProcess.stdin || !serverProcess.stdout || !serverProcess.stderr) {
    throw new Error('Failed to start MCP server process');
  }

  // Set up process handling
  let serverReady = false;
  let testCompleted = false;

  serverProcess.stderr.on('data', (data) => {
    const message = data.toString();
    console.log(`[SERVER] ${message.trim()}`);

    if (message.includes('MCP server connected via stdio')) {
      serverReady = true;
      console.log('✅ Server is ready');
    }
  });

  serverProcess.on('error', (error) => {
    console.error(`❌ Server process error: ${error}`);
    if (!testCompleted) {
      process.exit(1);
    }
  });

  serverProcess.on('exit', (code) => {
    console.log(`Server process exited with code: ${code}`);
    if (!testCompleted && code !== 0) {
      console.error('❌ Server exited unexpectedly');
      process.exit(1);
    }
  });

  // Wait for server to be ready
  await new Promise<void>((resolve, reject) => {
    const timeout = setTimeout(() => {
      reject(new Error('Server startup timeout'));
    }, 10000);

    const checkReady = () => {
      if (serverReady) {
        clearTimeout(timeout);
        resolve();
      } else {
        setTimeout(checkReady, 100);
      }
    };
    checkReady();
  });

  console.log('Testing MCP protocol communication...');

  // Test 1: Initialize
  const initRequest = {
    jsonrpc: '2.0',
    id: 1,
    method: 'initialize',
    params: {
      protocolVersion: '2024-11-05',
      capabilities: { roots: { listChanged: true }, sampling: {} },
      clientInfo: { name: 'test-client', version: '1.0.0' },
    },
  };

  console.log('Sending initialize request...');
  serverProcess.stdin.write(JSON.stringify(initRequest) + '\n');

  // Read initialization response
  const initResponse = await new Promise<any>((resolve, reject) => {
    const timeout = setTimeout(() => reject(new Error('Initialize timeout')), 5000);

    serverProcess.stdout.once('data', (data) => {
      clearTimeout(timeout);
      try {
        const response = JSON.parse(data.toString());
        resolve(response);
      } catch (error) {
        reject(new Error(`Invalid JSON response: ${error}`));
      }
    });
  });

  if (initResponse.result) {
    console.log('✅ Initialize successful');
    console.log(`   Server: ${initResponse.result.serverInfo?.name || 'Unknown'}`);
  } else {
    console.error('❌ Initialize failed:', initResponse.error);
    throw new Error('Initialize failed');
  }

  // Send initialized notification
  const initializedNotification = {
    jsonrpc: '2.0',
    method: 'notifications/initialized',
    params: {},
  };
  serverProcess.stdin.write(JSON.stringify(initializedNotification) + '\n');

  // Test 2: List tools
  const listToolsRequest = {
    jsonrpc: '2.0',
    id: 2,
    method: 'tools/list',
  };

  console.log('Sending list tools request...');
  serverProcess.stdin.write(JSON.stringify(listToolsRequest) + '\n');

  const toolsResponse = await new Promise<any>((resolve, reject) => {
    const timeout = setTimeout(() => reject(new Error('List tools timeout')), 5000);

    serverProcess.stdout.once('data', (data) => {
      clearTimeout(timeout);
      try {
        const response = JSON.parse(data.toString());
        resolve(response);
      } catch (error) {
        reject(new Error(`Invalid JSON response: ${error}`));
      }
    });
  });

  if (toolsResponse.result) {
    const tools = toolsResponse.result.tools || [];
    console.log(`✅ List tools successful - Found ${tools.length} tools`);
    console.log('   Available tools:', tools.map((t: any) => t.name).join(', '));
  } else {
    console.error('❌ List tools failed:', toolsResponse.error);
    throw new Error('List tools failed');
  }

  // Test 3: Call a simple tool
  const statsRequest = {
    jsonrpc: '2.0',
    id: 3,
    method: 'tools/call',
    params: {
      name: 'get_receiver_stats',
      arguments: { format: 'summary' },
    },
  };

  console.log('Sending get_receiver_stats request...');
  serverProcess.stdin.write(JSON.stringify(statsRequest) + '\n');

  const statsResponse = await new Promise<any>((resolve, reject) => {
    const timeout = setTimeout(() => reject(new Error('Get stats timeout')), 10000);

    serverProcess.stdout.once('data', (data) => {
      clearTimeout(timeout);
      try {
        const response = JSON.parse(data.toString());
        resolve(response);
      } catch (error) {
        reject(new Error(`Invalid JSON response: ${error}`));
      }
    });
  });

  if (statsResponse.result) {
    console.log('✅ Get receiver stats successful');
    console.log('   Response received with content');
  } else {
    console.error('❌ Get receiver stats failed:', statsResponse.error);
    throw new Error('Get receiver stats failed');
  }

  console.log('\n🎉 All tests passed! MCP server is working correctly.');
  console.log('✅ Remote connection established');
  console.log('✅ MCP protocol communication working');
  console.log('✅ Tool calls successful');

  testCompleted = true;

  // Clean up
  serverProcess.kill();

  // Wait for process to exit
  await new Promise<void>((resolve) => {
    serverProcess.on('exit', () => resolve());
    setTimeout(resolve, 1000); // Fallback timeout
  });
}

async function main(): Promise<void> {
  try {
    // Get connection details
    const { host: remoteHost, port: remotePort } = await getRemoteHostAndPort();

    // Run the test
    await testRemoteMcpServer(remoteHost, remotePort);
  } catch (error) {
    console.error(`Test failed: ${error}`);
    process.exit(1);
  }
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}
